### DISCUSSION TITLE
support for the so called "texture healing"?

### DISCUSSION THREAD

**Discussion by michaelssingh** [harfbuzz/harfbuzz#4490] [General] [answered]
Hello, GH released a pretty cool font face called [Monaspace](https://monaspace.githubnext.com/) which led me to wonder if Emacs supported such a thing as "texture healing", and thus I ended up here,  writing this issue (now discussion).

If so what would the correct term in harfbuzz be for "texture healing", how can it be enabled in Emacs?

**Comment by khaledhosny** [comment, marked as answer]
This seems to be regular OpenType substitution and positioning, but whether it works in Emacs or not is a question for Emacs.

**Reply by michaelssingh** [reply]
Just so we are clear,  GH refers to "texture healing " as:

> Texture healing preserves the monospace grid, and works in most editors without needing new software or editor plugins. How does it work?

**Comment by Eli-Zaretskii** [comment]
Emacs will support this as long as the underlying script is passed to HarfBuzz for shaping.  Plain ASCII text is generally *not* passed to HarfBuzz (because that would significantly slow down Emacs redisplay due to how text shaping is integrated into the Emacs display engine).  One can force Emacs to pass ASCII text to HarfBuzz by suitable customizations, but then don't come back complaining that redisplay became annoyingly slow.

**Comment by michaelssingh** [comment]
> Emacs will support this as long as the underlying script is passed to HarfBuzz for shaping. Plain ASCII text is generally _not_ passed to HarfBuzz (because that would significantly slow down Emacs redisplay due to how text shaping is integrated into the Emacs display engine). One can force Emacs to pass ASCII text to HarfBuzz by suitable customizations, but then don't come back complaining that redisplay became annoyingly slow.

Where exactly is the relevance of passing plain ASCII text in the case of Monaspace? Is that what fonts with ligatures are/do?

**Comment by Eli-Zaretskii** [comment]
Emacs does not use HarfBuzz unless the characters have entries in `composition-function-table`. Normally, only characters from scripts that require complex shaping and/or composition, such as Arabic and Hangul, have non-nil entries in that table. But you can define such entries for ASCII characters as well, and yes, this is one way of getting support for ligatures in ASCII text.

**Comment by stevemolitor** [comment]
I did get "textural healing" with Monaspace to work as described [here](https://github.com/mickeynp/ligature.el/issues/53#issuecomment-1828732077), although I tuned it for Commit Mono which uses a similar technique.

Eli forewarned me that this sort of thing could cause serious performance problems with redisplay. It's working for me, but be forewarned - it's not a recommended way of using the composition function table.
